cli/wallet/file/file.go

```diff
@@ -0,0 +1,348 @@
+	_, err := rand.Read(salt[:])
+	if err != nil {
+		return nil, err
+	}
+
+	envelope := &secretStateEnvelope{
+		Algorithm:  "pbkdf2",
+		Salt:       salt[:],
+		Iterations: 4096,
+	}
+	key, err := envelope.deriveKey(passphrase)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := json.Marshal(s)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize a Deoxys-II instance with the provided key and encrypt.
+	aead, err := deoxysii.New(key)
+	if err != nil {
+		return nil, err
+	}
+	envelope.Data = aead.Seal(nil, envelope.Salt[:aead.NonceSize()], data, nil)
+
+	return envelope, nil
+}
+
+type secretStateEnvelope struct {
+	Algorithm  string `json:"algorithm"`
+	Salt       []byte `json:"salt"`
+	Iterations uint32 `json:"iterations"`
+	Data       []byte `json:"data"`
+}
+
+func (e *secretStateEnvelope) deriveKey(passphrase string) ([]byte, error) {
+	switch e.Algorithm {
+	case "pbkdf2":
+		return pbkdf2.Key([]byte(passphrase), e.Salt, int(e.Iterations), stateKeySize, sha512.New), nil
+	default:
+		return nil, fmt.Errorf("unsupported key derivation algorithm: %s", e.Algorithm)
+	}
+}
+
+func (e *secretStateEnvelope) Open(passphrase string) (*secretState, error) {
+	// Derive key.
+	key, err := e.deriveKey(passphrase)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize a Deoxys-II instance with the provided key and decrypt.
+	aead, err := deoxysii.New(key)
+	if err != nil {
+		return nil, err
+	}
+	pt, err := aead.Open(nil, e.Salt[:aead.NonceSize()], e.Data, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Deserialize the inner state.
+	var state secretState
+	if err := json.Unmarshal(pt, &state); err != nil {
+		return nil, err
+	}
+
+	return &state, nil
+}
+
+func getWalletFilename(name string) string {
+	return filepath.Join(config.Directory(), fmt.Sprintf("%s.wallet", name))
+}
+
+type fileWalletFactory struct {
+	flags *flag.FlagSet
+}
+
+func (wf *fileWalletFactory) Kind() string {
+	return walletKind
+}
+
+func (wf *fileWalletFactory) Flags() *flag.FlagSet {
+	return wf.flags
+}
+
+func (wf *fileWalletFactory) GetConfigFromFlags() (map[string]interface{}, error) {
+	cfg := make(map[string]interface{})
+	cfg[cfgAlgorithm], _ = wf.flags.GetString(cfgAlgorithm)
+	return cfg, nil
+}
+
+func (wf *fileWalletFactory) RequiresPassphrase() bool {
+	// A file-backed wallet always requires a passphrase.
+	return true
+}
+
+func (wf *fileWalletFactory) unmarshalConfig(raw map[string]interface{}) (*walletConfig, error) {
+	if raw == nil {
+		return nil, fmt.Errorf("missing configuration")
+	}
+
+	var cfg walletConfig
+	if err := mapstructure.Decode(raw, &cfg); err != nil {
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+func (wf *fileWalletFactory) Create(name string, passphrase string, rawCfg map[string]interface{}) (wallet.Wallet, error) {
+	cfg, err := wf.unmarshalConfig(rawCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate entropy.
+	entropy, _ := bip39.NewEntropy(256)
+	mnemonic, _ := bip39.NewMnemonic(entropy)
```

Comment:
could we panic on these, if you'd rather not propagate the errors

the implementation of NewEntropy looks like it returns unrandomized bytes if rand.Read fails. we handle rand.Read failures elsewhere in this program.

Reply:
Yeah these should definitely propagate errors, not sure how this got missed.